File: WaTF-Bank-iOS/watf-bank/PageController/Favourite.m
#import "Util.h"
#import "SqlCipher.h"
#import "Http.h"
#import "TransferResult.h"

@interface Favourite:UIViewController <UIPickerViewDelegate,UIPickerViewDataSource,UITextFieldDelegate>
@property (weak, nonatomic) IBOutlet UITextField *amount;
@property (weak, nonatomic) IBOutlet UIPickerView *picker;
@end

@implementation Favourite

NSArray *pickerData;

-(void)viewDidLoad
{
    [super viewDidLoad];
    self.picker.dataSource = self;
    self.picker.delegate = self;
    pickerData = [SqlCipher selectAll];
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
{
    [_amount resignFirstResponder];
}

- (BOOL)textField:(UITextField *)textField shouldChangeCharactersInRange:(NSRange)range replacementString:(NSString *)string
{
    NSCharacterSet *numbersOnly = [NSCharacterSet characterSetWithCharactersInString:@"0123456789"];
    NSCharacterSet *characterSetFromTextField = [NSCharacterSet characterSetWithCharactersInString:string];
    BOOL stringIsValid = [numbersOnly isSupersetOfSet:characterSetFromTextField];
    return stringIsValid;
}


- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView
{
    return 1;
}

- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component
{
    return pickerData.count;
}

- (NSString*)pickerView:(UIPickerView *)pickerView titleForRow:(NSInteger)row forComponent:(NSInteger)component
{
    return pickerData[row];
}

- (IBAction)push:(id)sender {
    if([[SqlCipher selectAll] count]==0) return;
    NSString *accountNo = [SqlCipher select:[self pickerView:_picker titleForRow:[_picker selectedRowInComponent:0] forComponent:0]];
    if(accountNo==nil) return;
    NSDictionary *input = [NSDictionary dictionaryWithObjectsAndKeys:[Util getKeychain:@"accountNo"],@"accountNo",accountNo,@"toAccountNo",_amount.text,@"amount",[Util getKeychain:@"token"],@"token", nil];
    [Http post:@"transfer" :input :^(NSDictionary *res)
    {
        if(res==nil)
        {
            [Util alert:self :@"Connection Problem"];
            return;
        }
        if([[res[@"message"] lowercaseString] isEqualToString:@"success"])
        {
            TransferResult *t = [self.storyboard instantiateViewControllerWithIdentifier:@"TransferResult"];
            t.fromAccountV = [NSString stringWithFormat:@"%@ : %@",@"From",[Util getKeychain:@"accountNo"]];
            t.toAccountV = [NSString stringWithFormat:@"%@ : %@",@"To",[SqlCipher select:pickerData[[self.picker selectedRowInComponent:0]]]];
            t.amountV = [NSString stringWithFormat:@"%@ : %@",@"Amount",self.amount.text];
            [self presentViewController:t animated:true completion:nil];
        }
        else
        {
            [Util alert:self :@"An Error Occurred"];
        }
    }];
}

@end

File: iOS-Swift/watf-bank/PageController/Favourite.swift
import UIKit

class Favourite : UIViewController, UIPickerViewDelegate, UIPickerViewDataSource, UITextFieldDelegate {
    
    @IBOutlet weak var picker: UIPickerView!
    @IBOutlet weak var amount: UITextField!
    
    var pickerData = SQLCipher.selectAll("fav")
    
    override func viewDidLoad() {
        super.viewDidLoad()
        picker.delegate = self
        picker.dataSource = self
    }
    
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        amount.resignFirstResponder()
    }
    
    func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
        let allowChar = CharacterSet(charactersIn: "0123456789")
        return  allowChar.isSuperset(of: CharacterSet(charactersIn: string))
    }
    
    func numberOfComponents(in pickerView: UIPickerView) -> Int {
        return 1
    }
    
    func pickerView(_ pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
        return pickerData.count
    }
    
    func pickerView(_ pickerView: UIPickerView, titleForRow row: Int, forComponent component: Int) -> String? {
        return pickerData[row]
    }
    
    @IBAction func push(_ sender: Any) {
        if pickerData.count == 0 {
            Util.alert(self, "No Favourite")
            return
        }
        guard let token = KeyChain.load("token"), let acc = KeyChain.load("accountNo") else {
            Util.alert(self, "token")
            return
        }
        let toAcc = SQLCipher.selectFav(pickerView(picker, titleForRow: picker.selectedRow(inComponent: 0), forComponent: 0)!)[0]
        let input = ["accountNo":acc,"toAccountNo":toAcc,"amount":amount.text!,"token":token]
        Http().post(input, "transfer", completionHandler: {(re: [String:Any]) in
            var res = re as! [String:String]
            if let e = res["error"] {
                Util.alert(self, e)
            }
            else if res["message"] == "Success" {
                let t = self.storyboard!.instantiateViewController(withIdentifier: "TransferResult") as! TransferResult
                t.labelV = "Transfer Succeeded"
                t.fromAccountV = "From : "+res["username"]!
                t.toAccountV = "To : "+res["toAccount"]!
                t.amountV = "Amount : "+res["amount"]!
                self.present(t, animated: true, completion: nil)
            }
            else if let m = res["message"] {
                Util.alert(self, m)
            }
            else {
                Util.alert(self, "An Error Occurred")
            }
        })
    }
    
}
